src/modules/collection.ts

```diff
@@ -0,0 +1,200 @@
+import { isReadable } from '../utils/readable'
+import { BeeArgumentError, BeeError } from '../utils/error'
+
+const dirsEndpoint = '/dirs'
+const bzzEndpoint = '/bzz'
+
+interface CollectionUploadHeaders extends UploadHeaders {
+  'swarm-index-document'?: string
+  'swarm-error-document'?: string
+}
+
+export interface CollectionUploadOptions extends UploadOptions {
+  indexDocument?: string
+  errorDocument?: string
+}
+
+function extractCollectionUploadHeaders(options?: CollectionUploadOptions): CollectionUploadHeaders {
+  const headers: CollectionUploadHeaders = extractUploadHeaders(options)
+
+  if (options?.indexDocument) headers['swarm-index-document'] = options.indexDocument
+
+  if (options?.errorDocument) headers['swarm-error-document'] = options.errorDocument
+
+  return headers
+}
+
+function isUint8Array(obj: unknown): obj is Uint8Array {
+  return obj instanceof Uint8Array
+}
+
+function isCollection(data: unknown): data is Collection<Uint8Array | Readable> {
+  if (!Array.isArray(data)) {
+    return false
+  }
+
+  return !data.some(
+    entry =>
+      typeof entry !== 'object' || !entry.data || !entry.path || !(isUint8Array(entry.data) || isReadable(entry.data)),
+  )
+}
+
+function packData(data: Collection<Uint8Array | Readable>): TarArchive {
+  const tar = new TarArchive()
+
+  for (const entry of data) {
+    if (isUint8Array(entry.data)) {
+      tar.addUint8Array(entry.path, entry.data)
+    } else if (isReadable(entry.data)) {
+      if (entry.size === undefined) {
+        throw new BeeError(`Size has to be specified for Readable data! [${entry.path}]`)
+      }
+
+      tar.addStream(entry.path, entry.size, entry.data)
+    }
+  }
+
+  return tar
+}
+
+/**
+ * Creates array in the format of Collection with Readable streams prepared for upload.
+ *
+ * @param dir absolute path to the directory
+ * @param recursive flag that specifies if the directory should be recursively walked and get files in those directories.
+ */
+export async function buildCollection(dir: string, recursive = true): Promise<Collection<Readable>> {
+  // Handles case when the dir is not existing or it is a file ==> throws an error
+  const entries = await fs.promises.opendir(dir)
+  let collection: Collection<Readable> = []
+
+  for await (const entry of entries) {
+    const fullPath = path.join(dir, entry.name)
+
+    if (entry.isFile()) {
+      collection.push({
+        path: fullPath,
+        size: (await fs.promises.stat(fullPath)).size,
+        data: fs.createReadStream(fullPath),
+      })
+    } else if (entry.isDirectory() && recursive) {
+      collection = [...(await buildCollection(fullPath, recursive)), ...collection]
+    }
+  }
+
+  return collection
+}
+
+export async function buildFileListCollection(fileList: FileList): Promise<Collection<Uint8Array>> {
+  const collection: Collection<Uint8Array> = []
+
+  for (let i = 0; i < fileList.length; i++) {
+    const file = fileList.item(i)
+
+    if (file != null) {
```

Comment:
```suggestion
    if (file) {
```

I suggest we introduce eslint rule `no-eq-null` https://eslint.org/docs/rules/no-eq-null

Reply:
Good idea